src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1268,15 +1268,43 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         let opaque_defn_ty = tcx.type_of(opaque_def_id);
                         let opaque_defn_ty = opaque_defn_ty.subst(tcx, opaque_decl.substs);
                         let opaque_defn_ty = renumber::renumber_regions(infcx, &opaque_defn_ty);
+                        let concrete_is_opaque = infcx
```

Comment:
I'm not sure if this is quite the right check, or at least it doesn't match the behavior specified by the RFC. In the RFC, it's expected that the following would compile:

```rust
existential type Foo: Debug;

fn foo() -> Foo {
    5u32
}

fn bar(cond: bool) -> Foo {
    if cond { foo() } else { 5u32 }
}
```

In this situation, the type of the variable returned by `foo` is an impl trait variable, but since the function is within the same module as the definition of the `existential type`, it can constrain the type further and itself become a defining use. The RFC says you can even do this:

```rust
existential type Foo: Debug;

fn foo() -> Foo {
    5u32
}

fn bar() -> Foo {
    let x: u32 = foo();
    x + 5
}
```

I realize this isn't at all what's implemented today, but it'd be good to talk about how the behavior being implemented here interacts with the desired behavior we want to end up with.

Reply:
I believe that this check is still compatible with implementing the full behavior specified in the RFC. This check occurs after we've attempted to instantiate the opaque type. If we're still left with an opaque type after instantiation, then there's nothing else we can do.

Once the rest of the RFC is implemented, this check should be hit only when the opaque type is being used outside the defining module.

Reply:
That makes sense-- thanks for the clarification!